core/market/decentralized/src/matcher.rs

```diff
@@ -140,60 +181,334 @@ impl Matcher {
+                    .await
+                    // Some errors don't mean we shouldn't propagate unsubscription.
+                    .or_else(|e| match e {
+                        ModifyOfferError::UnsubscribedNotRemoved(..) => Ok(()),
+                        _ => Err(e),
+                    })
+                    // Collect Offers, that were correctly unsubscribed.
+                    .map(|_| offer_id.clone())
+                    .map_err(|e| match e {
+                        // We don't want to warn about normal situations.
+                        ModifyOfferError::Unsubscribed(..) | ModifyOfferError::Expired(..) => e,
+                        _ => {
+                            log::warn!(
+                                "Failed to unsubscribe Offer [{}]. Error: {}",
+                                &offer_id,
+                                &e
+                            );
+                            e
+                        }
+                    })
+                    .ok()
             }
-            ModifyOfferError::NotFound(_) => Ok(Propagate::No(Reason::NotFound)),
-            ModifyOfferError::Unsubscribed(_) => Ok(Propagate::No(Reason::Unsubscribed)),
-            ModifyOfferError::Expired(_) => Ok(Propagate::No(Reason::Expired)),
         })
+        .collect::<Vec<SubscriptionId>>()
+        .await;
+
+    if !new_unsubscribes.is_empty() {
+        log::info!(
+            "Received new Offers to unsubscribe from [{}]: \n{}",
+            caller,
+            DisplayVec(&new_unsubscribes)
+        );
+    }
+    Ok(new_unsubscribes)
+}
+
+// =========================================== //
+// Cyclic broadcasting
+// =========================================== //
+
+async fn random_broadcast_offers(matcher: Matcher) {
+    let broadcast_interval = matcher
+        .config
+        .clone()
+        .discovery
+        .mean_random_broadcast_interval
+        .to_std()
+        .map_err(|e| format!("Invalid broadcast interval. Error: {}", e))
+        .unwrap();
+    loop {
+        let matcher = matcher.clone();
+        async move {
+            let random_interval = randomize_interval(broadcast_interval);
+            tokio::time::delay_for(random_interval).await;
+
+            // We always broadcast our own Offers.
+            let our_offers = matcher
+                .list_our_offers()
+                .await?
+                .into_iter()
+                .map(|offer| offer.id)
+                .collect::<Vec<SubscriptionId>>();
+
+            // Add some random subset of Offers to broadcast.
+            let num_ours_offers = our_offers.len();
+            let num_to_broadcast = matcher.config.discovery.num_broadcasted_offers;
+
+            let all_offers = matcher
+                .store
+                .get_offers(None)
+                .await?
+                .into_iter()
+                .map(|offer| offer.id)
+                .collect::<Vec<SubscriptionId>>();
```

Comment:
Don't query full Offers from database if we only need ids

Reply:
Added comment for future

Reply:
fixed